Update to_python method in Elasticsearch adapters and refactor pillows to use them

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
The PR adds the `from_python` method in the ES adapters which is equivalent to `transform_.+_for_elasticsearch`. Now the adapter will have the ability to take in high-level python objects and make it suitable for indexing to ES.

After that, the PR updates the places where these `transform` functions are used. Pillows were the major consumer of these functions. So `ElasticProcessor` and `Reindexer` in pillows have been refactored to use the new adapter layer and simultaneously remove the transformation logic from the pillows.

This refactor lead to the complete removal of `ElasticSearchInterface` from pillows. Now all the calls to ES in HQ are going through the adapter layer. There was good test coverage on the 

Apart from that, all the tests have been modified and refactored that use the transformation functions. Only tests that equate transformation function and from_python function use `transform` function which will be removed in next PR with transformation functions.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14329
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The changes had good coverage and tests have been run locally.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-4881
<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
update user/signals.py to use native index instead of transform function

AmitPhulera committed:
make from_python and to_json instance methods

AmitPhulera committed:
add from_python in user_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
add from_python in domain_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python in app_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
add from_python in case_search_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
add from_python in case_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python in form_adapter

AmitPhulera committed:
add tests

AmitPhulera committed:
fixup domain implementation

AmitPhulera committed:
fix known failing tests

snopoke commented on corehq/apps/callcenter/tests/test_utils.py:
```diff
@@ -462,7 +462,7 @@ def test_is_midnight_for_domain(self):
 
 @es_test(requires=[domain_adapter])
-class CallCenterDomainTest(SimpleTestCase):
+class CallCenterDomainTest(TestCase):
```
why does this need DB access now?

AmitPhulera replied:
When we are trying to index domain, it is calling `form_python` which queries Subscriptions model.

snopoke replied:
That's unfortunate since those are optional fields

AmitPhulera replied:
I guess I can mock the call and change test back to `SimpleTestCase`.

snopoke reviewed (commented):
Overall this looks great. A few general questions:

* Why do some adapters not have a `to_dict` method?
* It seems like you've duplicated a lot of the code from the `transform_to_es` functions. Why not just use those in the adapter?

AmitPhulera commented:
> Why do some adapters not have a to_dict method?

I think only the app adapter doesn't have `from_dict` method. It was because the there was a single key that was added to the app object before saving it to ES. So it was pretty straightforward, so no `from_dict` was added for it.

> t seems like you've duplicated a lot of the code from the transform_to_es functions. Why not just use those in the adapter?

Yep, that is correct. As mentioned in the docstrings the functions are an almost exact replica of the transform functions. The transform functions will be removed in the subsequent commits and I am currently in process of updating pillows to not transform and let the adapter do the transformation at the end.

@snopoke

AmitPhulera committed:
add from_python in group_adapter

AmitPhulera committed:
add from_python in sms_adapter

AmitPhulera committed:
update ElasticProcessor to use adapter

- It removes the requirement of doc_prep_fn as it is handled by adapter

AmitPhulera committed:
update BulkElasticProcessor to use adapter

AmitPhulera committed:
update send_to_elasticsearch use the adapter and temporaily make all args optional

AmitPhulera committed:
isort + remove unused fns

AmitPhulera committed:
update case_pillow as per recent changes

AmitPhulera committed:
update domain_processor implementation

AmitPhulera committed:
update group es_processor defn

AmitPhulera committed:
update sms es_processor defn?

AmitPhulera committed:
update user es_processor

AmitPhulera committed:
update xform es_processor

AmitPhulera committed:
update application es_processor

AmitPhulera committed:
don't transform before sending doc to adapter

millerdev commented on corehq/apps/es/forms.py:
```diff
@@ -57,9 +69,92 @@ class ElasticForm(ElasticDocumentAdapter):
+        from corehq.pillows.utils import format_form_meta_for_es, get_user_type
+        from corehq.pillows.xform import is_valid_date
+
+        doc_ret = deepcopy(xform)
```
Is this one just too hard to eliminate `deepcopy`? In terms of runtime cost, it may still be worth it even if it's tedious.

AmitPhulera replied:
[4b9d19a](https://github.com/dimagi/commcare-hq/pull/32660/commits/4b9d19ae0333c1476dde2736f0cad8efe9139ee5)

millerdev reviewed (commented):
I didn't do a detailed review of every adapter here, but the comments on the user adapter should apply to all others as well.

AmitPhulera committed:
update get_checkpoint fn to take in index name instead of index info

AmitPhulera committed:
update references of get_checkpoint fn in pillows

AmitPhulera committed:
remove doc tranformation from build_bulk_payload

AmitPhulera committed:
update case_search es_processor defn

AmitPhulera committed:
update app_pillow test

AmitPhulera committed:
update case_search_pillow test

AmitPhulera committed:
update domain pillow tests

AmitPhulera committed:
update form pillow tests

AmitPhulera committed:
update user pillow tests

AmitPhulera committed:
update sms pillow tests

AmitPhulera committed:
update transform_xform_test to test to_json

AmitPhulera committed:
update xform pillow tests

AmitPhulera committed:
update user pillow tests

AmitPhulera committed:
update group pillow tests

AmitPhulera committed:
update case pillow tests

AmitPhulera committed:
update test basics tests

AmitPhulera committed:
make from_dict private

AmitPhulera committed:
optimize from_python in forms

AmitPhulera commented on corehq/apps/es/users.py:
```diff
@@ -73,9 +75,51 @@ class ElasticUser(ElasticDocumentAdapter):
+            raise TypeError(f"Unkown type {type(user)}")
+        return self.from_dict(user_dict)
+
+    def from_dict(self, user):
```
> what do you think of always passing in a dict that can be mutated, so the decision to copy or not can be pushed up to from_python which can make a more intelligent decision? For example, no copy is necessary when the thing being passed in here came from user.to_json().

The idea crossed my mind when I was implementing it, but I held back from doing it because we would be forced to use `deepcopy` everytime. Only that would ensure that the dict passed is safe to mutate. If we give this responsibility to `from_dict` then we could judicially copy it as per our requirements.

AmitPhulera replied:
Making `from_dict` private [55abccf](https://github.com/dimagi/commcare-hq/pull/32660/commits/55abccf9e5d256f1eac62dc740a540169bfcc5d0)

joxl replied:
> Making `from_dict` private [55abccf](https://github.com/dimagi/commcare-hq/pull/32660/commits/55abccf9e5d256f1eac62dc740a540169bfcc5d0)

+1, the private method is a nice change.

Going a bit further, it seems with the new `from_python()` flow, it wouldn't be such a stretch to just remove `_from_dict()` entirely and in-line the logic inside the `from_python()` method now (example below). This is just something that comes to mind from my (possibly naive) reviewer perspective, it's completely fine if you prefer to keep them separate.

```python
    def from_python(self, doc):
        from corehq.some.models import ModelClass
        if isinstance(doc, ModelClass):
            doc_dict = doc.to_json()
        elif isinstance(doc, dict):
            doc_dict = doc
        else:
            raise TypeError(f"Invalid type: {type(doc)} (expected {ModelClass} or dict)")

        # all code in `_from_dict()` can be added here, in-line

        return doc_dict.pop("_id"), doc_dict
```

AmitPhulera commented on corehq/apps/es/client.py:
```diff
@@ -431,19 +431,17 @@ def export_adapter(self):
                   indexed/updated/deleted in Elasticsearch
         """
-        raise NotImplementedError(f"{cls.__name__} is abstract")
+        raise NotImplementedError(f"{self.__name__} is abstract")
```
No, it does not. Thanks for catching it.

AmitPhulera replied:
[4e78bc3](https://github.com/dimagi/commcare-hq/pull/32660/commits/4e78bc384656ca63b5406373e0779a41c2fed13e)

AmitPhulera committed:
fix error message bug

AmitPhulera committed:
Update corehq/apps/es/tests/test_user_adapter.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

millerdev commented on corehq/apps/es/forms.py:
```diff
@@ -92,42 +92,46 @@ def _from_dict(cls, xform):
         doc_ret['user_type'] = get_user_type(user_id)
         doc_ret['inserted_at'] = datetime.utcnow().isoformat()
 
+        form['meta'] = form_meta
+        doc_ret['form'] = form
+
```
See other comment about copies.
```suggestion
```

millerdev commented on corehq/apps/es/forms.py:
```diff
@@ -92,42 +92,46 @@ def _from_dict(cls, xform):
         from corehq.pillows.utils import format_form_meta_for_es, get_user_type
         from corehq.pillows.xform import is_valid_date
+        doc_ret = copy(xform)
+        form = copy(doc_ret['form'])
+        form_meta = copy(form.get('meta', None))
```
Assigning the copies immediately here would make it easier to see how they relate to `doc_ret` eventually.
```suggestion
        form = doc_ret['form'] = copy(doc_ret['form'])
        form_meta = form['meta'] = copy(form.get('meta', None))
```

millerdev commented on corehq/apps/es/users.py:
```diff
@@ -73,9 +75,51 @@ class ElasticUser(ElasticDocumentAdapter):
+        if isinstance(user, CouchUser):
+            user_dict = user.to_json()
+        elif isinstance(user, dict):
+            user_dict = user
```
Copying discussion context from [other comment](https://github.com/dimagi/commcare-hq/pull/32660#discussion_r1116686135):

> > what do you think of always passing in a dict that can be mutated, so the decision to copy or not can be pushed up to `from_python` which can make a more intelligent decision? For example, no copy is necessary when the thing being passed in here came from `user.to_json()`.
> 
> The idea crossed my mind when I was implementing it, but I held back from doing it because we would be forced to use `deepcopy` everytime. Only that would ensure that the dict passed is safe to mutate. If we give this responsibility to `from_dict` then we could judicially copy it as per our requirements.

Are you still intending to implement this, or did I miss the reason we don't want to push `copy` up to `from_python` now that `_from_dict` is private?

AmitPhulera replied:
I have a different thought process here-
We have two scenarios, first when we want `from_dict` the ability to alter the dicts that are being passed in. In that case we would have to deepcopy the dicts and pass it to `from_dict`
The second case is not allowing `from_dict` to mutate the dict which is happening right now. In this case `from_dict` either copies or deepcopies based on the requirements and never alters the dicts.

If we push `copy` to `from_python` it will create a scenario where we sometimes would be mutating the dict and sometime we will not be. I am not sure if it the right way to go.

AmitPhulera commented on corehq/apps/es/forms.py:
```diff
@@ -92,42 +92,46 @@ def _from_dict(cls, xform):
         from corehq.apps.receiverwrapper.util import get_app_version_info
         from corehq.pillows.utils import format_form_meta_for_es, get_user_type
         from corehq.pillows.xform import is_valid_date
+        doc_ret = copy(xform)
```
`_from_dict` is taking in a variable name `xform` so this should work fine. But yeah renaming that to `xfomr_dict` would be better.

AmitPhulera committed:
assign the copies earlier of newly created dicts

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
remove unused fn

AmitPhulera committed:
refactor es_utils to not use ElasticSearchInterface

AmitPhulera committed:
correct params in assume_alias calls

AmitPhulera committed:
correct params in mapping_exists calls

AmitPhulera committed:
correct params in initialize_index fn calls

AmitPhulera committed:
correct params in set_index_normal_settings calls

AmitPhulera committed:
correct params in set_index_reindex_settings calls

AmitPhulera committed:
add fn to get index_info from adapter

AmitPhulera committed:
refactor ElasticPillowReindexer to use adapter and update related functions

AmitPhulera committed:
correct params for modified fns

AmitPhulera committed:
update ResumableBulkElasticPillowReindexer and BulkPillowReindexProcessor to accept adapter and remove ElasticSearchInterface related args

AmitPhulera committed:
update all pillows with correct args ResumableBulkElasticPillowReindexer

AmitPhulera committed:
Merge branch 'ap/es/update-to-python' of https://github.com/dimagi/commcare-hq into ap/es/update-to-python

AmitPhulera committed:
update _to_dict in form adapter to cover some edge cases

AmitPhulera committed:
update test_bulk as per new implementaion

AmitPhulera committed:
remove tests that were equation ESInterface args with adapter args

AmitPhulera committed:
fix tests with patching correct fns

AmitPhulera committed:
update test_elasticsearch to use adapter

AmitPhulera committed:
remove interface.py

AmitPhulera committed:
remvoe last reference of ElasticSearchInterface

AmitPhulera committed:
fix other known test failures

AmitPhulera committed:
mock db call instead of switchin to TestCase

AmitPhulera committed:
merge commit

joxl commented on corehq/apps/es/transient_util.py:
```diff
@@ -73,6 +73,19 @@ def index_info_from_cname(cname):
     return CANONICAL_NAME_INFO_MAP[cname]
 
 
+def index_info_from_adapter(adapter):
```
Adding a new utility to map from an adapter back to an index info object feels a bit backwards. How about changing the functions that still need info objects to support both? This causes an issue with the `assume_alias()` function because adapters know nothing about index aliases, but that is fine -- creating aliases on the Elasticsearch cluster itself can safely be dropped since aliases haven't been used for anything other than adapter lookups in code for quite some time now.

See 689c434715f94ecdab91769ccc78ccd5bc8b94a3 for a full proposal.

millerdev replied:
@AmitPhulera Is this suggestion expected to be implemented in this PR?

AmitPhulera replied:
I looked into it and supporting both would require bringing back support for ES Info objects in the ESProcesser. 
So in this PR I plan to keep it as it is. I have added a TODO in the code which I plan to do with the final cleanups that are planned after this PR is merged.

joxl reviewed (changes requested):
The only blocking changes are `from_python()` methods that can return `None` for the document id. All other comments are non-blocking.

AmitPhulera commented on corehq/apps/users/signals.py:
```diff
@@ -29,13 +29,7 @@ def _update_user_in_es(couch_user):
-        user_adapter.index(_transform_for_es(couch_user))
-
-
-def _transform_for_es(couch_user):
```
This is already done but not pushed 😅 
I have been grappling with one test that I haven’t figured out yet. I am thinking it sorted first and then push all the changes together.

AmitPhulera commented on corehq/apps/es/apps.py:
```diff
@@ -32,9 +36,29 @@ class ElasticApp(ElasticDocumentAdapter):
+            raise TypeError(f"Unknown type {type(app)}")
+        app_obj['@indexed_on'] = json_format_datetime(datetime.utcnow())
+        app_dict = app_obj.to_json()
+        return app_dict.pop('_id', None), app_dict
```
Yep yep. It totally makes sense. I will update the functions.

AmitPhulera committed:
remove tranform_to_es functions out of tests

AmitPhulera committed:
patch db calls in tests where required

AmitPhulera committed:
correct patch path

AmitPhulera commented on corehq/pillows/ledger.py:
```diff
@@ -99,9 +99,9 @@ def get_ledger_to_elasticsearch_pillow(pillow_id='LedgerToElasticsearchPillow',
     """
     assert pillow_id == 'LedgerToElasticsearchPillow', 'Pillow ID is not allowed to change'
-    IndexInfo = namedtuple('IndexInfo', ['index'])
+    index_name = "ledgers_2016-03-15"
```
Intersting...Maybe? 🤔 . I will ask team about it. Even if it is being used, it should be a no-op change

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/es/update-to-python

joxl commented on testapps/test_pillowtop/tests/test_xform_pillow.py:
```diff
@@ -247,12 +243,13 @@ def test_transform_xform_for_elasticsearch_app_versions(self):
             }
         }
-        doc_ret = transform_xform_for_elasticsearch(doc_dict)
+        doc_ret = form_adapter.to_json(doc_dict)
```
That sounds right. No need to do it in this PR.

AmitPhulera committed:
fix more tests

AmitPhulera committed:
fix doc warnings

AmitPhulera committed:
don't return None when _id is not present

AmitPhulera committed:
pass a copy of model dict in _from_python

AmitPhulera committed:
remove all leftover transform functions from code

AmitPhulera committed:
remove transform fns from tests

now only place that uses transform fns are tests that check equality of transform fns and to_python of adapter

AmitPhulera committed:
correct patch path

AmitPhulera committed:
add TODO for removal of a fn

AmitPhulera committed:
fix docstrings

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/es/update-to-python

AmitPhulera commented on custom/covid/tests/test_custom_rules.py:
```diff
@@ -107,7 +103,7 @@ def test_custom_action(self):
         for case in [patient_case, other_patient_case, other_case]:
             case_search_adapter.index(
-                transform_case_for_elasticsearch(case.to_json()),
+                case.to_json(),
```
https://github.com/dimagi/commcare-hq/pull/32660/commits/bd4c3bd55201ba2c0831b068298a964768d79ef2#diff-92c28b7322f70ef5e77e14567e6060f8a8ac5b3c2404746ac71ee35958d8d33cR106

calellowitz commented on corehq/apps/es/domains.py:
```diff
@@ -55,9 +60,38 @@ class ElasticDomain(ElasticDocumentAdapter):
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, domain):
```
given that this is basically the same in all adapters, I think it would make sense to break it out into a util function (maybe one that takes the expected class as a parameter), or even just on the base class and have the doc class part of the subclass

AmitPhulera replied:
That is a good suggestion. I thought about it when I was implementing it but given how this function is used which is majorly in index function, the API of an index would be changed to take in a class also which I guess would not be ideal. So I did not go that route.

snopoke replied:
I don't quite follow your reasoning @AmitPhulera. If you made it a function you could just call the function here instead of having the full implementation.

If you did want to put it on the base class you could still call it from here with the correct type (or else add the type as a class level attribute so that it doesn't need to be passed in.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/0b0c9914bf4a6affe39035ee23efef4eacd55456 I have moved the function to base class and subsequent commits updates the usage in adapters.

calellowitz commented on corehq/apps/es/domains.py:
```diff
@@ -55,9 +60,38 @@ class ElasticDomain(ElasticDocumentAdapter):
+            raise TypeError(f"Unkown type {type(domain)}")
+        return self.from_dict(domain_dict)
+
+    def from_dict(self, domain_dict):
```
Why is this sometimes public and sometimes private?

AmitPhulera replied:
Thanks for catching it. It was missed when I was making this fn private in all the adapters.
Fixed in https://github.com/dimagi/commcare-hq/pull/32660/commits/e939243489fdacb4882ec4932cde054c48eca7d7

calellowitz reviewed (approved):
I did a pretty high level review without getting too deep into the weeds and no blocking comments, but had a few questions

AmitPhulera committed:
merge master - fix conflicts

AmitPhulera committed:
make from_dict private in leftover domain adapter

snopoke commented on corehq/apps/es/domains.py:
```diff
@@ -55,9 +60,38 @@ class ElasticDomain(ElasticDocumentAdapter):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, domain):
+        """
+        :param user: an instance of ``CouchUser`` or a user dict
```
```suggestion
        :param user: an instance of ``Domain`` or a domain dict
```

snopoke commented on corehq/apps/es/apps.py:
```diff
@@ -32,9 +36,29 @@ class ElasticApp(ElasticDocumentAdapter):
+        """
+        Takes in an ``Application`` object or an app dict
+        and applies required transformation to make it suitable for ES.
+        The function is replica of ``transform_app_for_es`` with added support for user objects.
```
> with added support for user objects.

what does this mean?

snopoke commented on corehq/apps/export/tests/test_get_export_file.py:
```diff
@@ -676,27 +676,27 @@ class ExportTest(SimpleTestCase):
+                name='batman',
+                case_json={"foo": "apple", "bar": "banana", "date": '2016-4-24'},
+            )
+            case_adapter.index(case.to_json(), refresh=True)
```
nit: remove unnecessary `to_json` call
```suggestion
            case_adapter.index(case, refresh=True)
```

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/db1087b79ae0d50322410862443f03e2a20eeb21

snopoke commented on corehq/apps/export/tests/test_get_export_file.py:
```diff
@@ -676,27 +676,27 @@ class ExportTest(SimpleTestCase):
+                owner_id="some_other_owner",
+                case_json={"foo": "apple", "bar": "banana", "date": '2016-4-04'},
+            )
+            case_adapter.index(case.to_json(), refresh=True)
```
```suggestion
            case_adapter.index(case, refresh=True)
```

snopoke replied:
nit: you could use `case_adapter.bulk_index`

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/db1087b79ae0d50322410862443f03e2a20eeb21

snopoke commented on corehq/apps/export/tests/test_get_export_file.py:
```diff
@@ -676,27 +676,27 @@ class ExportTest(SimpleTestCase):
-        case = new_case(type="some_other_type", case_json={"foo": "apple", "bar": "banana"})
-        case_adapter.index(case.to_json(), refresh=True)
+            case = new_case(type="some_other_type", case_json={"foo": "apple", "bar": "banana"})
+            case_adapter.index(case.to_json(), refresh=True)
```
```suggestion
            case_adapter.index(case, refresh=True)
```

snopoke commented on corehq/apps/export/tests/test_get_export_file.py:
```diff
@@ -676,27 +676,27 @@ class ExportTest(SimpleTestCase):
-        case = new_case(closed=True, case_json={"foo": "apple", "bar": "banana"})
-        case_adapter.index(case.to_json(), refresh=True)
+            case = new_case(closed=True, case_json={"foo": "apple", "bar": "banana"})
+            case_adapter.index(case.to_json(), refresh=True)
```
```suggestion
            case_adapter.index(case, refresh=True)
```

snopoke commented on corehq/apps/hqadmin/tests/test_stale_data_in_es.py:
```diff
@@ -244,22 +242,17 @@ def _submit_form(self, domain, new_cases=0, update_cases=()):
-                FormDocumentStore(form.domain, form.xmlns).get_document(form.form_id)
-            )
+            es_form = FormDocumentStore(form.domain, form.xmlns).get_document(form.form_id)
             form_adapter.index(es_form, refresh=True)
```
nit: could use `bulk_index`

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/ca5fdb0bb3d8ba3c6f19041f8c8097ea75156bda

snopoke commented on corehq/ex-submodules/pillowtop/processors/elastic.py:
```diff
@@ -97,26 +93,19 @@ def process_change(self, change):
                 return
 
-            # prepare doc for es
-            doc_ready_to_save = self.doc_transform_fn(doc)
```
This will change the datadog timing data since the 'transform' step now happens under 'load'.

I suspect that's not and issue and we could potentially remove these metrics @calellowitz

snopoke commented on corehq/ex-submodules/pillowtop/processors/elastic.py:
```diff
@@ -189,42 +175,40 @@ def process_changes_chunk(self, changes_chunk):
 
-def send_to_elasticsearch(index_info, doc_type, doc_id, es_getter, name, data=None,
-                          delete=False, es_merge_update=False):
+def send_to_elasticsearch(index_info=None, doc_type=None, doc_id=None, es_getter=None, name=None, adapter=None,
```
Making all the args optional is a bit dangerous since it's not clear what combination of args to provide.

I would suggest creating separate functions with different signatures which call a private function under the hood.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/028e281a5a0a081e1254ed65b34a622b4de340b1 I did a quick search around in codebase and I realised that its usage has already been refactored. So I removed the params that are no longer needed and marked the essential args non-optional.

snopoke commented on corehq/ex-submodules/pillowtop/reindexer/reindexer.py:
```diff
@@ -144,19 +145,21 @@ def reindex(self):
+def prepare_index_for_reindex(adapter):
+    es = get_client()
+    index_info = index_info_from_adapter(adapter)
     initialize_index_and_mapping(es, index_info)
```
the `es` param is unused so can be set to None and remove the `es` var above
```suggestion
    initialize_index_and_mapping(None, index_info)
```

AmitPhulera replied:
This I will do when I work on removing es param totally.

snopoke commented on corehq/apps/api/tests/case_resources.py:
```diff
@@ -127,11 +125,11 @@ def test_parent_and_child_cases(self):
-            transform_case_for_elasticsearch(child_case.to_json()),
+            child_case,
             refresh=True
         )
```
nit: could use `bulk_index`

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/ca5fdb0bb3d8ba3c6f19041f8c8097ea75156bda

snopoke commented on corehq/apps/data_interfaces/tests/test_case_deduplication.py:
```diff
@@ -49,10 +47,7 @@ def setUp(self):
-                transform_case_for_elasticsearch(case.to_json()),
-                refresh=True
-            )
+            case_search_adapter.index(case, refresh=True)
```
nit: could use `bulk_index`

snopoke replied:
Also other places in this test

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/ca5fdb0bb3d8ba3c6f19041f8c8097ea75156bda

snopoke commented on corehq/apps/es/apps.py:
```diff
@@ -32,9 +36,29 @@ class ElasticApp(ElasticDocumentAdapter):
+            app_obj = app
+        else:
+            raise TypeError(f"Unknown type {type(app)}")
+        app_obj['@indexed_on'] = json_format_datetime(datetime.utcnow())
```
seems odd to do this on the object and not on the dict version of it

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/11e06b400c99fc7d1f93a42a1cfd21f0e73d07c5

snopoke commented on corehq/apps/es/cases.py:
```diff
@@ -53,9 +56,37 @@ class ElasticCase(ElasticDocumentAdapter):
+                case["owner_id"] = case["user_id"]
+
+        case['owner_type'] = get_user_type(case.get("owner_id", None))
+        case['inserted_at'] = datetime.utcnow().isoformat()
```
pity this isn't called `@indexed_on` :(

snopoke replied:
I know you just copied this but it should probably use `json_format_datetime`

AmitPhulera replied:
> pity this isn't called @indexed_on :(

Do you think it would be a good idea to add `@indexed_on` as a separate key and during reindex update `inserted_at` with `@indexed_on`?

> I know you just copied this but it should probably use json_format_datetime

Does `json_format_datetime` is timezone aware and add appends `Z` to denote UTC, wouldn't it result in having different `datetime` string formats for the same key? Should it be considered a problem or do you think its fine to update the formats? 

@snopoke

snopoke replied:
> > pity this isn't called @indexed_on :(
> 
> Do you think it would be a good idea to add `@indexed_on` as a separate key and during reindex update `inserted_at` with `@indexed_on`?

It would be nice for consistency but we'd also then need to make changes to APIs that reference the `inserted_at` so it would be a larger task and I'm not sure it's worth it.


> > I know you just copied this but it should probably use json_format_datetime
> 
> Does `json_format_datetime` is timezone aware and add appends `Z` to denote UTC, wouldn't it result in having different `datetime` string formats for the same key? Should it be considered a problem or do you think its fine to update the formats?

Technically I think it's an improvement since it's now clear that the date is in UTC. Under the hood ES converts dates to UTC for storage and any date without a TZ is assumed to be UTC so I don't think it would cause any issue.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/c660f9b6bd7f7f54eb22e8ad1e68bdec219ccaec

snopoke commented on corehq/apps/es/forms.py:
```diff
@@ -57,9 +69,100 @@ class ElasticForm(ElasticDocumentAdapter):
+                pass
+
+        xform_dict['user_type'] = get_user_type(user_id)
+        xform_dict['inserted_at'] = datetime.utcnow().isoformat()
```
I know you just copied this but it should probably use `json_format_datetime`

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/32660/commits/c660f9b6bd7f7f54eb22e8ad1e68bdec219ccaec

snopoke commented on corehq/ex-submodules/pillowtop/es_utils.py:
```diff
@@ -45,28 +45,29 @@ def to_json(self):
 
 def initialize_index_and_mapping(es, index_info):
-    index_exists = es.indices.exists(index_info.index)
+    # TODO: Remove es param as it is unused
```
will this be done separately?

AmitPhulera replied:
Yep. Tests are a heavy consumer of this function. I have thought of a followup PR in which more cleanup would be done which will resolve this TODO.

snopoke commented on corehq/ex-submodules/pillowtop/processors/elastic.py:
```diff
@@ -127,7 +116,7 @@ def _datadog_timing(self, step):
             tags={
                 'action': step,
-                'index': self.index_info.alias,
+                'index': self.adapter.index_name,  # TODO: try to add index cname
```
> TODO: try to add index cname

using the actual index name is probably more useful since you could tell if a new index is slower than an old index

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/es/update-to-python

AmitPhulera committed:
add from_python and _from_dict in base class

AmitPhulera committed:
update user adapter class

AmitPhulera committed:
update case_search adapter class

AmitPhulera committed:
update app adapter class

AmitPhulera committed:
update case adapter class

AmitPhulera committed:
update domain adapter class

AmitPhulera committed:
update form adapter class

AmitPhulera committed:
update groups and sms adapter class

AmitPhulera committed:
add docstrings for model_cls variable

AmitPhulera committed:
update send_to_elasticsearch to make adapter and make some args compulsory

AmitPhulera committed:
bulk index case docs in test_get_export_file.py

AmitPhulera committed:
remove redundant to_json calls on model objects in tests

AmitPhulera committed:
remove todo

AmitPhulera committed:
use bulk_index whenever possible

millerdev reviewed (approved):
Looking great!

AmitPhulera committed:
replace intialize_index_and_mappings with @es_test in tests

AmitPhulera committed:
update more tests to use @es_test for index setup

AmitPhulera committed:
remove unused util test fns

AmitPhulera committed:
remove requirement of low level es client object in initialize_index_and_mapping

AmitPhulera committed:
update app_adapter to use base class fns

AmitPhulera committed:
use json_format_datetime

AmitPhulera commented on corehq/apps/es/apps.py:
```diff
@@ -56,9 +56,9 @@ def from_python(self, app):
-        app_obj['@indexed_on'] = json_format_datetime(datetime.utcnow())
         app_dict = app_obj.to_json()
-        return app_dict.pop('_id'), app_dict
+        app_dict['@indexed_on'] = json_format_datetime(datetime.utcnow())
```
https://github.com/dimagi/commcare-hq/pull/32660/commits/d59dacfa4ae71b3d83ed275e60a69f2ba0bfd677
